src/cocaine-app/balancer.py

```diff
@@ -1598,16 +1601,57 @@ def storage_keys_diff(self, request):
         return {'couples': couples_diff,
                 'total_keys_diff': sum(couples_diff.itervalues())}
 
-    @h.source
-    @h.handler_wne
-    def get_cached_keys(self, request):
+    def _update_cached_keys(self):
+        start_ts = time.time()
+        logger.info('Cached keys updating: started')
+        try:
+            self._do_update_cached_keys()
+        except Exception as e:
+            logger.exception('Cached keys updating: failed: {}'.format(e))
+            self._cached_keys = e
+        finally:
+            logger.info('Ceched keys updating: finished, time: {0:.3f}'.format(
+                time.time() - start_ts))
+            self.__tq.add_task_at(
+                'cached_keys_update',
+                self.cached_keys_timer.next(),
+                self._update_cached_keys)
+
+    @h.concurrent_handler
+    def force_update_cached_keys(self, request):
+        start_ts = time.time()
+        logger.info('Cached keys forced updating: started')
+        try:
+            self._do_update_cached_keys()
+        except Exception as e:
+            logger.exception('Cached keys forced updating: failed: {}'.format(e))
+            self._cached_keys = e
+        finally:
+            logger.info('Cached keys forced updating: finished, time: {0:.3f}'.format(
+                time.time() - start_ts))
+
+    def _do_update_cached_keys(self):
         if not config.get("cache") or not config.get('metadata', {}).get('cache', {}):
-            yield {}
-            raise StopIteration
+            self._cached_keys = {}
+            return
         mc = ReconnectableService(
             '{base_name}-cache'.format(base_name=config.get('app_name', 'mastermind')),
             attempts=3, timeout=10, logger=logger)
-        yield mc.enqueue('get_cached_keys', msgpack.packb(None))
+        try:
+            self._cached_keys = mc.enqueue('get_cached_keys', msgpack.packb(None))
+        except Exception as e:
+            logger.exception('Cached keys updating: failed: {}'.format(e))
```

Comment:
You don't need to embed exception into a log message, 'logger.exception' does it by itself.